File: src/ServerScriptService/AppStart/GameLoop.server.lua
local ServerScriptService = game:GetService("ServerScriptService")

local gameModulesFolder = ServerScriptService:WaitForChild("GameModules")
local GameManager = require(gameModulesFolder:WaitForChild("GameManager", 1))

local appStartFolder = ServerScriptService:WaitForChild("AppStart", 1)
local Spawners = require(appStartFolder:WaitForChild("Spawners", 1))
local npcAgent = require(ServerScriptService:WaitForChild("Agents"):WaitForChild("NpcAgent"))

--- Main injection point for the game
-- @script ServerMain

local require = require(game:GetService("ReplicatedStorage"):WaitForChild("Nevermore"))

local function OneTimeSetup()
    Spawners.Init()
    GameManager:Initialize()
end

local function RunForever()
    while true do
        repeat
            GameManager:RunIntermission()
        until GameManager:GameReady()
        
        GameManager:StopIntermission()
        GameManager:StartRound()
        npcAgent.CreateMaleRunner()
        npcAgent.CreateFemaleRunner()
        repeat
            GameManager:Update()
            wait(0.1)
        until GameManager:RoundOver()
        
        GameManager:RoundCleanup()
    end
end

OneTimeSetup()
RunForever()

File: src/ServerScriptService/AppStart/GameLoop.server.ts
import { ServerScriptService, ReplicatedStorage } from "@rbxts/services";
// Run this before the other stuff
const nevermoreModule = ReplicatedStorage.WaitForChild("Nevermore") as ModuleScript;
const nevermoreInitialize = require(nevermoreModule);
// Nevermore initialized, safe to proceed
import { TestRunner } from './StartupTests/TestRunner';
import { IGameManager } from '../GameModules/GameModulesTypings';

import { GameJector } from './GameJector';

const GameManager = GameJector.GetDefaultGame() as IGameManager

function PreFlightCheck() : void {
    if (!TestRunner.RunTests()) {
        throw "Failed startup tests"
    }
}

function OneTimeSetup() : void {
    GameManager.Initialize()
    GameManager.BeforeGameStart()
}

function RunForever() : void {
    while (true) {
        while (!GameManager.GameReady()) {
            GameManager.RunIntermission();
            wait(0.2);
        }
        GameManager.BeforeRoundStart()
        GameManager.StopIntermission();
        GameManager.StartRound();
        
        while (!GameManager.RoundOver()) {
            GameManager.Update();
            wait(0.1);
        }

        GameManager.RunIntermission()
        GameManager.RoundCleanup()
    }
}

PreFlightCheck();
OneTimeSetup();
RunForever();
